v2/component/launcher.go

```diff
@@ -242,92 +253,259 @@ func NewLauncher(runtimeInfo string, options *LauncherOptions) (*Launcher, error
+func (l *Launcher) storeOutputParameterValueFromCache(cachedExecution *pb.Execution) (*metadata.Parameters, error) {
+	mlmdOutputParameters, err := cacheutils.GetOutputParamsFromCachedExecution(cachedExecution)
+	if err != nil {
+		return nil, err
+	}
+	// Read output parameters.
+	outputParameters := &metadata.Parameters{
+		IntParameters:    make(map[string]int64),
+		StringParameters: make(map[string]string),
+		DoubleParameters: make(map[string]float64),
+	}
+
+	for name, param := range l.runtimeInfo.OutputParameters {
+		filename := param.Path
+		outputParamValue, ok := mlmdOutputParameters[name]
+		if !ok {
+			return nil, fmt.Errorf("can't find parameter %v in mlmdOutputParameters", name)
+		}
+		if err := ioutil.WriteFile(filename, []byte(outputParamValue), 0644); err != nil {
+			return nil, fmt.Errorf("failed to write output parameter %q to file %q: %w", name, filename, err)
+		}
+		switch param.Type {
+		case "STRING":
+			outputParameters.StringParameters[name] = outputParamValue
+		case "INT":
+			i, err := strconv.ParseInt(strings.TrimSpace(outputParamValue), 10, 0)
 			if err != nil {
-				return fmt.Errorf("unable to parse input artifact id from %q: %w", id, err)
+				return nil, fmt.Errorf("failed to parse parameter name=%q value =%v to int: %w", name, outputParamValue, err)
 			}
-			ecfg.InputArtifactIDs[name] = append(ecfg.InputArtifactIDs[name], id)
+			outputParameters.IntParameters[name] = i
+		case "DOUBLE":
+			f, err := strconv.ParseFloat(strings.TrimSpace(outputParamValue), 0)
+			return nil, fmt.Errorf("failed to parse parameter name=%q value =%v to double: %w", name, outputParamValue, err)
+			outputParameters.DoubleParameters[name] = f
+		default:
+			return nil, fmt.Errorf("unknown type. Expected STRING, INT or DOUBLE")
 		}
 	}
+	return outputParameters, nil
+}
 
-	for name, parameter := range executorInput.Inputs.Parameters {
-		switch t := parameter.Value.(type) {
-		case *pipeline_spec.Value_StringValue:
-			ecfg.InputParameters.StringParameters[name] = parameter.GetStringValue()
-		case *pipeline_spec.Value_IntValue:
-			ecfg.InputParameters.IntParameters[name] = parameter.GetIntValue()
-		case *pipeline_spec.Value_DoubleValue:
-			ecfg.InputParameters.DoubleParameters[name] = parameter.GetDoubleValue()
-		default:
-			return fmt.Errorf("unknown parameter type: %T", t)
+func (l *Launcher) storeOutputArtifactMetadataFromCache(ctx context.Context, executorInputOutputs *pipeline_spec.ExecutorInput_Outputs, cachedMLMDExecutionID int64) ([]*metadata.OutputArtifact, error) {
+	MLMDOutputArtifacts, err := l.metadataClient.GetOutputArtifactsByExecutionId(ctx, cachedMLMDExecutionID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get MLMDOutputArtifacts by executionId %v: %w", cachedMLMDExecutionID, err)
+	}
+	MLMDOutputArtifactByName := make(map[string]*pb.Artifact)
+	for _, artifact := range MLMDOutputArtifacts {
+		name := extractNameFromURI(*artifact.Uri)
```

Comment:
Name should be taken from the event that connects artifact and the original execution.
See event.path

Reply:
No. I printed the event.path. It records the id of the artifact which will change every time.

Reply:
@capri-xiyue strange, did you check event.artifactId or event.path?
Refer to code https://github.com/kubeflow/pipelines/blob/ad419cd63dc9a6c7bc109c509311bdec5f671e57/v2/metadata/client.go#L236

Reply:
I printed the wrong one. Fixed it in the refactor PR.